Meter.h

```diff
@@ -20,7 +20,8 @@ in the source distribution for its full text.
 
 
 #define METER_TXTBUFFER_LEN 256
-#define METER_GRAPHDATA_SIZE 256
+#define MIN_METER_GRAPHDATA_RECORDS 128
+#define MAX_METER_GRAPHDATA_RECORDS 2048
```

Comment:
Why would you use htop with such a tiny font size? This is meant to protect against allocating a useless and super large buffer. Guarding against integer overflow would be a bonus here. 2048 records are large - with each record being one second, 2048 records would take 34 minutes to fill them all.

Reply:
> Why would you use htop with such a tiny font size?

Different people have different preferences?  And as I mentioned, screen size is a factor too (and those do tend to get larger over time).

> This is meant to protect against allocating a useless and super large buffer.

If htop is running in a big enough terminal, it's _not_ useless, it's necessary for correct behavior.

> Guarding against integer overflow would be a bonus here.

Is there a particular calculation you're concerned with potentially overflowing?  If so I think that might be better addressed at the point of the calculation, not hand-wavily upstream of it.

> 2048 records are large - with each record being one second, 2048 records would take 34 minutes to fill them all.

From the system I'm typing this on:

```
$ ps -o lstart,cmd -p 7923
                 STARTED CMD
Wed Feb  9 16:16:18 2022 htop -d20
```

I routinely leave htop running for very long periods of time, and it's often useful to be able to look back at what various metrics were in the recent past (that's the whole point of the graph display, no?).

Reply:
@zevweiss The limit is still needed, but I now raised it up to `(UINT_MAX/sizeof(*data->values))`. Would this be a good limit?

Reply:
> I now raised it up to (UINT_MAX/sizeof(*data->values)). Would this be a good limit?

So I gather you want to cap the array at `UINT_MAX` bytes?  I don't see why we'd choose that particular value, but that's not really my main concern, because ultimately...

> The limit is still needed

I disagree -- any number we choose is going to be arbitrary, and I don't think we should generally be in the business of imposing arbitrary limits.

Reply:
> I disagree -- any number we choose is going to be arbitrary, and I don't think we should generally be in the business of imposing arbitrary limits.

It's more of a limit imposed by the Curses environment and I don't choose this arbitrarily. The Curses library cannot have a terminal width larger than `INT_MAX`, and the maximum buffer size for `malloc` is `SIZE_MAX`. Rather than writing the limit like:

```
MININUM(INT_MAX, SIZE_MAX/sizeof(*data->values))
```

... I would choose a limit that's slightly smaller, but sufficient to avoid the integer overflow.

Reply:
Okay, let's crunch some numbers and see if any of this makes sense.

Let's be conservative and assume a 64K UltraHiRes display with 65536 horizontal pixels and a font of 4 pixels for each horizontal character (which is already barely readable as is).

If we would now open a htop screen in this hell of an unreadable display, we'd be presented with close to 16384 columns, which is half of what curses would allow us to do. Not to mention an unpatched X11 server would have given in at 2048 horizontal pixels, just FYI (more modern ones are AFAIR limited at 8K, but that won't change the result much either).

Or phrased differently: Just make the limit 32768 and be done with it as any attempt to display that full history buffer is likely to be limited by being unable to read it or by means of the display environment crashing on you before you reach that resolution.

On the other hand, let's assume a 16 bit platform trying to run htop: Good luck getting the main binary loaded in the first place. No need to worry about those either.

Lastly: Just make the number of elements allocated in the array `size_t`. It will be good enough and the actual memory layout is no real concern for the main data structures.

Reply:
@BenBE I don't know what you are suggesting. I've already raised the limit to above even 32768. And the limit is now only there to prevent the `unsigned int` overflow.
Besides the ncurses limit (terminal width has type `int`), I use a powerOf2Ceil function and I just don't want to extend that function to work with `size_t` as it's just not necessarily for this case.

Reply:
Scrap the `powerOf2Ceil`/`powerOf2Floor` stuff as this (as of now) adds far too much noise for the scope of this patch and instead stick to the MIN/MAX approach by @zevweiss with `MIN` set to 128 and `MAX` set at 32768 (reasoning for these numbers above).

Reply:
While I can comply and change my code to remove the power-of-2 rounding, I don't think @zevweiss fully understood what the upper limit is for.

Zev's approach was to pick the larger of `w * 2` and 1.5 × `data->numRecords`. While it's good enough, it's vulnerable to an arithmetic overflow when calculating the new size. My main concern is on 32-bit systems, where changing the `unsigned int` type to `size_t` doesn't help preventing the overflow as the two types are both 32-bits. The real way to prevent it is to cap the maximum of `data->numRecords` to `SIZE_MAX/sizeof(*data->values)` or smaller. Since I have to make a limit anyway, I'm facing a choice of using a limit that's as large as the system allowed, or a smaller, "arbitrary" as he said, but reasonable value like 32768. It seems like BenBE suggested me for the latter.